serialization.rst

```diff
@@ -0,0 +1,337 @@
+
+  - A BSON array will be deserialized as a PHP ``array``.
+  - A BSON document (root or nested) without ``__pclass`` property [1]_ becomes a
+    PHP ``stdClass`` object, with each BSON document property becoming a
+    public ``stdClass`` property.
+  - A BSON document (root or nested) with ``__pclass`` property [1]_ becomes
+    a PHP object of the class name as defined by the ``__pclass`` property.
+
+    If the named class implements the ``MongoDB\BSON\Unserializable``
+    interface, then the properties of the BSON document, minus the
+    ``__pclass`` property are sent as an associative array to the
+    ``bsonUnserialize`` function to initialise the object's properties.
+    
+    If the named class does not implement the ``MongoDB\BSON\Unserializable``
+    interface, then the object gets initialized with the properties
+    **without** the ``__pclass`` property being one of them. It simply gets
+    left out.
+
+    The rationale for the latter is because no explicit choice about whether
+    to use ODM features has been made, it should not bail out because of an
+    extra ``__pclass`` property. The behaviour is different if the data type
+    was mapped to ``"odm"`` (see below).
+
+- ``"array"`` — turns a BSON array or BSON document into a PHP array.
+  ``__pclass`` properties [1]_ are igored.
+
+- ``"object"`` or ``"stdClass"`` — turns a BSON array or BSON document into a
+  ``stdClass`` object. There will be no special treatment of a ``__pclass``
+  property [1]_, buy it should **not** be set as property in the returned
+  object.
+
+- ``"odm"`` — uses the class name set in the ``__pclass`` property of a BSON
+  array or BSON document to determine the class name of the created object.
+
+  The properties of the BSON document, minus the ``__pclass`` property are
+  sent as an associative array to the ``bsonUnserialize`` function to initialise
+  the object's properties.
+
+  Remarks:
+
+  - If the ``__pclass`` property is not present [1]_, this throws the
+    ``MongoDB\Driver\Exception\UnexpectedValueException`` exception with a
+    message indicating the special class marker could not be found.
+
+  - If the class as determined by the ``__pclass`` property does not implement
+    the ``MongoDB\BSON\Persistable`` interface, throw an
+    ``MongoDB\Driver\Exception\UnexpectedValueException`` exception with a
+    message indicating that the named class does not implement the required
+    interface.
+
+- ``any other string`` — defines the class name that the BSON array or BSON
+  object should be deserialized at.
+
+  If the class implements the ``MongoDB\BSON\Unserializable`` interface, then
+  the properties of the BSON document, **minus** the ``__pclass`` property [1]_
+  if it exists, are sent as an associative array to the ``bsonUnserialize``
+  function to initialise the object's properties.
+
+  To discuss:
+
+  - What should we do if the named class implements
+    ``MongoDB\BSON\Persistable``?
```

Comment:
For context, this is the case where the user specified a particular class name instead of "odm" and that class does may or may not match what we find in `__pclass`, or `__pclass` might not even exist.

It's possible that we're reading old data and the class only recently started implementing `Persistable`. Either way, I think we can proceed with calling `bsonUnserialize()` if the class implements the appropriate interface.

Consider a case where the specified class is `Persistable` and the `__pclass` value we found didn't match. This could be because the user is storing a class hierarchy in a single collection and decided that they want the document hydrated into the base class (e.g. BaseUser instead of AdminUser or ClientUser). Even thought the raw document is an AdminUser and has some extra fields, we're going to invoke `BaseUser::bsonUnserialize()` and let it decide what fields it cares about. I don't think that's a problem.

Obviously, if the class we infer (either a directly specified class name or for the "odm" mode and respecting `__pclass`) doesn't implement our `Unserializable` class, we have a problem and should throw.